### DISCUSSION TITLE
Reinstalling Budibase

### DISCUSSION THREAD

**Discussion by PaulMcF1987** [Budibase/budibase#5939] [Help] [answered]
I got removed Budibase in favour of another app builder, however I think the 2 could work well together...

I have tried reinstalling budibase on docker but  get:

`budibase_couch-init_1           /entrypoint.sh sh -c sleep ...   Exit 2`

I have checked the logs and they return
   
`  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current`
 `                                Dload  Upload   Total   Spent    Left  Speed`
``
`  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0`
`100    95  100    95    0     0   2495      0 --:--:-- --:--:-- --:--:--  2567`
`{"error":"file_exists","reason":"The database could not be created, the file already exists."}`
`  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current`
`                                 Dload  Upload   Total   Spent    Left  Speed`
`{"error":"file_exists","reason":"The database could not be created, the file already exists."}`
``
`  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0`
`100    95  100    95    0     0  58139      0 --:--:-- --:--:-- --:--:-- 95000`
`sh: fg: line 1: no current job`

It looks to me as though the database from my previous install is still there and preventing me creating one for this install.
I dont know my previous details to rejoin the old database.

does anyine know how to overcome this?

currently when I try to access Budibase I get:

![image](https://user-images.githubusercontent.com/75368133/168864331-972bb434-e4de-40b1-bd4b-00c225b30594.png)

**Comment by melohagan** [comment]
Hi @PaulMcF1987 

I would get the same warning message whenever I do a `docker-compose up` but I am able to access the builder successfully. 
It just means it doesn't create the database because it had already been created in the volume.

Are there any other errors, and are all of the other containers definitely up and running? 
<img width="401" alt="Screenshot 2022-05-18 at 11 27 53" src="https://user-images.githubusercontent.com/101575380/169018473-77f9d041-e1f2-49f2-ae99-8b1c4c6bfaa4.png">

Where are you hosting?

----
If you want to completely nuke your Budibase installation, then you can do a `docker volume ls` to see the list of volumes e.g.
![Screenshot 2022-05-18 at 11 25 45](https://user-images.githubusercontent.com/101575380/169018096-207c6c70-7700-4224-b680-2578b18987e0.png)

And then you can do a `docker volume rm budibase_couchdb3_data`.
**This will delete all of the apps, data, users, etc, from your budibase installation so only do this if you want a completely clean slate**

**Comment by PaulMcF1987** [comment, marked as answer]
@melohagan 
I am hosting on a VPS from Ionos.

All containers are running bar `budibase_couch-init_` which shows `Exit 2`

`docker volume ls ` gives

![image](https://user-images.githubusercontent.com/75368133/169019396-c7a85bf7-66ca-493e-8b4b-89c8edcc3eb6.png)

I tried running `docker volume rm budibase_couchdb3_data` and got the error  `Error response from daemon: remove budibase_couchdb3_data: volume is in use - [4e29ee1657459dc1898085cb787de7f787a2af441cdb3938565d4ccc31f8862b]`

**Reply by melohagan** [reply]
You probably need to stop and remove the budibase container first, but I would delete the volume as a last resort. 

Do a `docker ps -a` to see the status of the different services.

**Reply by PaulMcF1987** [reply]
@melohagan 
I ran `docker-compose down` and then `remove budibase_couchdb3_data` (I was using data from a different database so this database was not used to host any of my data)

But I still cannot get access.. 

Still getting `budibase_couch-init_1           /entrypoint.sh sh -c sleep ...   Exit 2 `
All other services are running

**Reply by melohagan** [reply]
Thanks. If you look at the logs for the **bbproxy**, and **bbapps** are there any errors in there?

**Reply by PaulMcF1987** [reply]
YML
```
`version: "3"`

`# optional ports are specified throughout for more advanced use cases.`

`services:`
`  app-service:`
`    restart: unless-stopped`
`    image: budibase.docker.scarf.sh/budibase/apps`
`    container_name: bbapps`
`    environment:`
`      SELF_HOSTED: 1`
`      COUCH_DB_URL: http://${COUCH_DB_USER}:${COUCH_DB_PASSWORD}@couchdb-service:5984`
`      WORKER_URL: http://worker-service:4003`
`      MINIO_URL: http://minio-service:9000`
`      MINIO_ACCESS_KEY: ${MINIO_ACCESS_KEY}`
`      MINIO_SECRET_KEY: ${MINIO_SECRET_KEY}`
`      INTERNAL_API_KEY: ${INTERNAL_API_KEY}`
`      BUDIBASE_ENVIRONMENT: ${BUDIBASE_ENVIRONMENT}`
`      PORT: 4002`
`      JWT_SECRET: ${JWT_SECRET}`
`      LOG_LEVEL: info`
`      SENTRY_DSN: https://a34ae347621946bf8acded18e5b7d4b8@o420233.ingest.sentry.io/5338131`
`      ENABLE_ANALYTICS: "true"`
`      REDIS_URL: redis-service:6379`
`      REDIS_PASSWORD: ${REDIS_PASSWORD}`
`    depends_on:`
`      - worker-service`
`      - redis-service`

`  worker-service:`
`    restart: unless-stopped`
`    image: budibase.docker.scarf.sh/budibase/worker`
`    container_name: bbworker`
`    environment:`
`      SELF_HOSTED: 1`
`      PORT: 4003`
`      CLUSTER_PORT: ${MAIN_PORT}`
`      JWT_SECRET: ${JWT_SECRET}`
`      MINIO_ACCESS_KEY: ${MINIO_ACCESS_KEY}`
`      MINIO_SECRET_KEY: ${MINIO_SECRET_KEY}`
`      MINIO_URL: http://minio-service:9000`
`      APPS_URL: http://app-service:4002`
`      COUCH_DB_USERNAME: ${COUCH_DB_USER}`
`      COUCH_DB_PASSWORD: ${COUCH_DB_PASSWORD}`
`      COUCH_DB_URL: http://${COUCH_DB_USER}:${COUCH_DB_PASSWORD}@couchdb-service:5984`
`      SENTRY_DSN: https://a34ae347621946bf8acded18e5b7d4b8@o420233.ingest.sentry.io/5338131`
`      INTERNAL_API_KEY: ${INTERNAL_API_KEY}`
`      REDIS_URL: redis-service:6379`
`      REDIS_PASSWORD: ${REDIS_PASSWORD}`
`    depends_on:`
`      - redis-service`
`      - minio-service`
`      - couch-init`

`  minio-service:`
`    restart: unless-stopped`
`    image: minio/minio`
`    volumes:`
`      - minio_data:/data`
`    environment:`
`      MINIO_ACCESS_KEY: ${MINIO_ACCESS_KEY}`
`      MINIO_SECRET_KEY: ${MINIO_SECRET_KEY}`
`      MINIO_BROWSER: "off"`
`    command: server /data`
`    healthcheck:`
`      test: ["CMD", "curl", "-f", "http://localhost:9000/minio/health/live"]`
`      interval: 30s`
`      timeout: 20s`
`      retries: 3`

`  proxy-service:`
`    restart: unless-stopped`
`    ports:`
`      - "${MAIN_PORT}:10000"`
`    container_name: bbproxy`
`    image: budibase/proxy`
`    depends_on:`
`      - minio-service`
`      - worker-service`
`      - app-service`
`      - couchdb-service`

`  couchdb-service:`
`    restart: unless-stopped`
`    image: ibmcom/couchdb3`
`    environment:`
`      - COUCHDB_PASSWORD=${COUCH_DB_PASSWORD}`
`      - COUCHDB_USER=${COUCH_DB_USER}`
`    volumes:`
`      - couchdb3_data:/opt/couchdb/data`

`  couch-init:`
`    image: curlimages/curl`
`    environment:`
`      PUT_CALL: "curl -u ${COUCH_DB_USER}:${COUCH_DB_PASSWORD} -X PUT couchdb-service:5984"`
`    depends_on:`
`      - couchdb-service`
`    command: ["sh","-c","sleep 10 && $${PUT_CALL}/_users && $${PUT_CALL}/_replicator; fg;"]`

`  redis-service:`
`    restart: unless-stopped`
`    image: redis`
`    command: redis-server --requirepass ${REDIS_PASSWORD}`
`    volumes:`
`      - redis_data:/data`

`  watchtower-service:`
`    restart: always`
`    image: containrrr/watchtower`
`    volumes:`
`      - /var/run/docker.sock:/var/run/docker.sock`
`    command: --debug --http-api-update bbapps bbworker bbproxy`
`    environment:`
`      - WATCHTOWER_HTTP_API=true`
`      - WATCHTOWER_HTTP_API_TOKEN=budibase`
`      - WATCHTOWER_CLEANUP=true`
`    labels:`
`      - "com.centurylinklabs.watchtower.enable=false"`

`volumes:`
`  couchdb3_data:`
`    driver: local`
`  minio_data:`
`    driver: local`
`  redis_data:`
`    driver: local`
```
ENV
```
`# Use the main port in the builder for your self hosting URL, e.g. localhost:10000`
`MAIN_PORT=10000`

`# This section contains all secrets pertaining to the system
`# These should be updated
`JWT_SECRET=Secret`
`MINIO_ACCESS_KEY=Access'
`MINIO_SECRET_KEY=Secret`
`COUCH_DB_PASSWORD=PASSWORD`
`COUCH_DB_USER=User`
`REDIS_PASSWORD=PASSWORD`
`INTERNAL_API_KEY=API`

`# This section contains variables that do not need to be altered under normal circumstances`
`APP_PORT=9900`
`WORKER_PORT=9901`
`MINIO_PORT=9902`
`COUCH_DB_PORT=9903`
`REDIS_PORT=9904`
`WATCHTOWER_PORT=9905`
`BUDIBASE_ENVIRONMENT=PRODUCTION`
```

**Reply by PaulMcF1987** [reply]
@melohagan 
BBapps Logs
```
TypeError [ERR_INVALID_URL]: Invalid URL: http://%IvXQVTN/global-db/
    at new NodeError (internal/errors.js:322:7)
    at onParseError (internal/url.js:270:9)
    at new URL (internal/url.js:346:5)
    at parseURL (/app/node_modules/node-fetch/lib/index.js:1164:12)
    at new Request (/app/node_modules/node-fetch/lib/index.js:1210:17)
    at /app/node_modules/node-fetch/lib/index.js:1439:19
    at new Promise (<anonymous>)
    at fetch (/app/node_modules/node-fetch/lib/index.js:1437:9)
    at fetchCookie (/app/node_modules/fetch-cookie/index.js:29:23)
    at processTicksAndRejections (internal/process/task_queues.js:95:5) {
  input: 'http://%IvXQVTN/global-db/',
  code: 'ERR_INVALID_URL'
}
[Tenant: default] [Migration: user_email_view_casing] [DB: global-db] Error:  TypeError: Cannot read property 'user_email_view_casing' of undefined
    at /app/node_modules/@budibase/backend-core/src/migrations/index.js:55:16
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async exports.doWithDB (/app/node_modules/@budibase/backend-core/src/db/index.js:59:12)
    at async runMigration (/app/node_modules/@budibase/backend-core/src/migrations/index.js:50:5)
    at async internal (/app/node_modules/@budibase/backend-core/src/context/index.js:90:14)
    at async /app/node_modules/@budibase/backend-core/src/context/FunctionContext.js:23:22
    at async exports.runMigrations (/app/node_modules/@budibase/backend-core/src/migrations/index.js:113:7)
Error performing migrations. Exiting.
 TypeError: Cannot read property 'user_email_view_casing' of undefined
    at /app/node_modules/@budibase/backend-core/src/migrations/index.js:55:16
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async exports.doWithDB (/app/node_modules/@budibase/backend-core/src/db/index.js:59:12)
    at async runMigration (/app/node_modules/@budibase/backend-core/src/migrations/index.js:50:5)
    at async internal (/app/node_modules/@budibase/backend-core/src/context/index.js:90:14)
    at async /app/node_modules/@budibase/backend-core/src/context/FunctionContext.js:23:22
    at async exports.runMigrations (/app/node_modules/@budibase/backend-core/src/migrations/index.js:113:7)
Close
```

BBProxy
```
2022/05/18 10:41:46 [error] 31#31: *13 connect() failed (111: Connection refused) while connecting to upstream, client: 82.25.245.114, server: _, request: "GET / HTTP/1.1", upstream: "http://172.30.0.8:4002/", host: "domainIP:10000", referrer: "http://domainIP:10000/"
2022/05/18 10:42:45 [info] 32#32: *15 client timed out (110: Connection timed out) while waiting for request, client: 82.25.245.114, server: 0.0.0.0:10000
````

**Reply by melohagan** [reply]
@PaulMcF1987 For the environment variables, did you replace the actual user and password for security?
```
`COUCH_DB_PASSWORD=PASSWORD`
`COUCH_DB_USER=User`
```
If you did, obviously don't share, but maybe there's some strange characters in the user or password that's causing problems. 
This URL is the problem: `http://%IvXQVTN/global-db/`
It should take the form: 
```
http://${COUCH_DB_USER}:${COUCH_DB_PASSWORD}@couchdb-service:5984/global-db
or
http://couchdb-service:5984/global-db
```

**Reply by PaulMcF1987** [reply]
@melohagan I did change the values of these. Are there any particular characters that I should have avoided?

**Reply by melohagan** [reply]
Here's the code that parses the URL: https://github.com/Budibase/budibase/blob/869a4fa91cd0444a19d5cc0eb9d2f9790ffc5c5a/packages/backend-core/src/db/pouch.js#L4

So I'd try avoiding ':', '/', and '@' and see if that helps. Or even as a test remove any special characters.

**Reply by PaulMcF1987** [reply]
@melohagan 
Problem solved, I removed the `@` from the password. Thank you for your help!

**Reply by melohagan** [reply]
Awesome! That's a nasty bug; I've raised it here: https://github.com/Budibase/budibase/issues/5962
